src/dataflow/src/render/reduce.rs

```diff
@@ -144,94 +142,111 @@ where
-                            // TODO(frank): Arguably, the absence of one aggregate is evidence
-                            // that the key doesn't exist (the others could be phantoms due to
-                            // negative input records); we could just suppress the output in that
-                            // case, rather than panic, though we surely want to see what is up.
-                            // XXX: This panic reports user-supplied data!
-                            panic!(
-                                "[customer-data] ReduceCollation found unexpected indexes:\n\tExpected:\t{:?}\n\tFound:\t{:?}\n\tFor:\t{:?}\n\tKey:{:?}\nRelationExpr:\n{}",
-                                (0..aggregates_len).collect::<Vec<_>>(),
-                                input.iter().map(|((p,_),_)| p).collect::<Vec<_>>(),
-                                aggregates_clone,
-                                key,
-                                relation_expr_clone.pretty(),
-                            );
-                        }
-                        row_packer.extend(key.iter());
-                        for ((_pos, val), cnt) in input.iter() {
-                            assert_eq!(*cnt, 1);
-                            row_packer.push(val.unpack_first());
-                        }
-                        output.push((row_packer.finish_and_reuse(), 1));
-                    }
-                });
-                (oks, err_input)
-            };
-            let index = (0..group_key.len()).collect::<Vec<_>>();
-            self.set_local_columns(relation_expr, &index[..], (oks, errs.arrange()));
+                    let is_accumulable = aggregates
+                        .iter()
+                        .map(|a| accumulable_hierarchical(&a.func).0)
+                        .collect::<Vec<_>>();
+                    differential_dataflow::collection::concatenate(scope, to_collect)
+                        .reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceCollation", {
+                            let mut row_packer = RowPacker::new();
+                            move |key, mut input, output| {
+                                // The inputs are pairs of an optional index and row to decode.
+
+                                // There can be at most one `None` index, and it indicates the accumulable aggregates.
+                                let new_row = Row::new(Vec::new());
+                                let mut accumulable = if (input[0].0).0 == None {
+                                    let iter = (input[0].0).1.iter();
+                                    input = &input[1..];
+                                    iter
+                                } else {
+                                    new_row.iter()
+                                };
+
+                                row_packer.extend(key.iter());
+                                for is_accum in is_accumulable.iter() {
```

Comment:
this whole loop is a bit silly because we have bunches of nicely packed data in various places and it seems like we should be able to just memcpy over what we want but instead we spend a bunch of time unpacking and repacking stuff